src/librustc_typeck/check/wfcheck.rs

```diff
@@ -253,16 +253,55 @@ impl<'ccx, 'gcx> CheckTypeWellFormedVisitor<'ccx, 'gcx> {
         });
     }
 
+    fn check_auto_trait(&mut self,
+                                   trait_def_id: DefId,
+                                   span: Span)
+    {
+        let predicates = self.tcx().lookup_predicates(trait_def_id);
+
+        // If we must exclude the Self : Trait predicate contained by all
+        // traits.
+        let no_refl_predicates : Vec<_> =
```

Comment:
So currently adding a constraint like that fails because Option requires a Sized typed, and adding a Sized bound to Magic fails, because there is a new predicate in the predicate set. I agree we should test the self type here, but I'm not sure the best way to capture this in a test.